### PR TITLE
set weights per layer

### PR DESCRIPTION
**Pull request by jenniew**
## What changes were proposed in this pull request?

allow set weight and bias per layer

## How was this patch tested?
unit tests

## Related links or issues (optional)
fixed https://github.com/intel-analytics/BigDL/issues/724

### CODE REVIEW

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/abstractnn/AbstractModule.scala [hunk 1]
```diff
@@ -345,5 +345,26 @@ abstract class AbstractModule[A <: Activity: ClassTag, B <: Activity: ClassTag,
...
     Predictor(this).predictClass(dataset)
   }
+
+  /**
+    * Set weight and bias for the module
+    * @param newWeights
+    * @return true: set weights successfully
+    *         false: set weights failed
+    */
+  def setWeights(newWeights: Array[Tensor[T]]): Boolean = {
```
Comment (qiuxin2012) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/abstractnn/AbstractModule.scala]: no return value

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/abstractnn/AbstractModule.scala [hunk 2]
```diff
@@ -345,5 +345,26 @@ abstract class AbstractModule[A <: Activity: ClassTag, B <: Activity: ClassTag,
...
+    if (parameters() == null) {
+      false
+    } else {
+      val weights = parameters()._1
+      // set weight
```
Comment (qiuxin2012) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/abstractnn/AbstractModule.scala]: The first one won't always be a weight. parameters() may return (weight), (weight, bias) and (bias).
 Just add a require(weights.length == newWeights.length), then use a while loop to copy the weights.

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/python/api/PythonBigDL.scala [hunk 3]
```diff
@@ -1294,6 +1294,20 @@ class PythonBigDL[T: ClassTag](implicit ev: TensorNumeric[T]) extends Serializab
   def initEngine(): Unit = {
     Engine.init
   }
+  
+  def setWeights(model: AbstractModule[Activity, Activity, T], weights: JList[JTensor]): Boolean = {
+    val weightTensor = weights.asScala.toArray.map(toTensor(_))
+    return model.setWeights(weightTensor)
+  }
+
+  def getWeights(model: AbstractModule[Activity, Activity, T]): JList[JTensor] = {
```
Comment (qiuxin2012) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/python/api/PythonBigDL.scala]: Please also add getWeights in abstractmodule.

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/abstractnn/AbstractModule.scala (hunk 1 above)
Comment (yiheng) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/abstractnn/AbstractModule.scala]: the name is confusing. -> setWeightsBias?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/abstractnn/AbstractModule.scala [hunk 4]
```diff
@@ -345,5 +345,26 @@ abstract class AbstractModule[A <: Activity: ClassTag, B <: Activity: ClassTag,
...
   }
+
+  /**
+    * Set weight and bias for the module
+    * @param newWeights
+    * @return true: set weights successfully
+    *         false: set weights failed
+    */
+  def setWeights(newWeights: Array[Tensor[T]]): Boolean = {
+    if (parameters() == null) {
```
Comment (yiheng) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/abstractnn/AbstractModule.scala]: why not throw an exception? If user input a wrong parameter, they will be confused if you swallow the error.

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/abstractnn/AbstractModule.scala [hunk 5]
```diff
@@ -345,5 +345,26 @@ abstract class AbstractModule[A <: Activity: ClassTag, B <: Activity: ClassTag,
...
+      false
+    } else {
+      val weights = parameters()._1
+      // set weight
+      weights(0).copy(newWeights(0))
```
Comment (yiheng) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/abstractnn/AbstractModule.scala]: This is not right for container

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/abstractnn/AbstractModule.scala (hunk 1 above)
Comment (yiheng) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/abstractnn/AbstractModule.scala]: I think return type should be this.type

Comment (jenniew) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/abstractnn/AbstractModule.scala]: This is consistant with keras. The keras api "set_weights" is passing a list of numpy arrays which includes weight and bias.

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/abstractnn/AbstractModule.scala (hunk 5 above)
Comment (qiuxin2012) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/abstractnn/AbstractModule.scala]: +1, you should override set/getWeights in Container. @jenniew

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/abstractnn/AbstractModule.scala (hunk 1 above)
Comment (jenniew) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/abstractnn/AbstractModule.scala]: fixed

Comment (jenniew) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/abstractnn/AbstractModule.scala]: Fixed

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/abstractnn/AbstractModule.scala (hunk 2 above)
Comment (jenniew) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/abstractnn/AbstractModule.scala]: Fixed

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/abstractnn/AbstractModule.scala (hunk 4 above)
Comment (jenniew) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/abstractnn/AbstractModule.scala]: Fixed

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/abstractnn/AbstractModule.scala (hunk 1 above)
Comment (jenniew) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/abstractnn/AbstractModule.scala]: Fixed

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/python/api/PythonBigDL.scala (hunk 3 above)
Comment (jenniew) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/python/api/PythonBigDL.scala]: Fixed

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/abstractnn/AbstractModule.scala [hunk 6]
```diff
@@ -345,5 +345,31 @@ abstract class AbstractModule[A <: Activity: ClassTag, B <: Activity: ClassTag,
...
     Predictor(this).predictClass(dataset)
   }
+
+  /**
+    * Set weight and bias for the module
+    * @param newWeights
+    * @return
+    */
+  def setWeights(newWeights: Array[Tensor[T]]): Unit = {
+    require(parameters() != null)
```
Comment (jenniew) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/abstractnn/AbstractModule.scala]: Fixed

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/abstractnn/AbstractModule.scala [hunk 7]
```diff
@@ -345,5 +345,31 @@ abstract class AbstractModule[A <: Activity: ClassTag, B <: Activity: ClassTag,
...
+    require(parameters()._1.length == newWeights.length)
+    val weights = parameters()._1
+    for(i <- 0 to newWeights.length-1) {
```
Comment (jenniew) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/abstractnn/AbstractModule.scala]: Fixed

**Comment by yiheng** [comment]
The branch cannot be rebased due to conflicts. Can you please fix this?

**Pull request merged**